### PR TITLE
Fixed #27792 -- Extended category support for feeds

### PR DESCRIPTION
**Pull request by PavloKapyshin**
[Ticket #27792](https://code.djangoproject.com/ticket/27792)

### CODE REVIEW

File: docs/ref/utils.txt [hunk 1]
```diff
@@ -347,13 +347,25 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
         ``self.feed``.
 
         All parameters should be strings, except ``categories``, which should
-        be a sequence of strings.
+        be a sequence of strings or instances of :class:`Category`.
+
+        .. versionchanged:: 2.0
+
+            Now it is possible to specify :class:`Category` instances
```
Comment (foresmac) [inline docs/ref/utils.txt]: Instead of "Now it is possible", try "It is now possible".

File: docs/ref/utils.txt [hunk 2]
```diff
@@ -347,13 +347,25 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
...
+            in ``categories`` argument.
```
Comment (foresmac) [inline docs/ref/utils.txt]: try "in the `categories` argument"

File: docs/ref/utils.txt [hunk 3]
```diff
@@ -347,13 +347,25 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
...
         ``pubdate`` and ``updateddate``, which are ``datetime.datetime``
-        objects, and ``enclosures``, which is a list of ``Enclosure`` instances.
+        objects; ``enclosures``, which is a list of ``Enclosure`` instances;
+        ``categories``, which is a sequence of strings or :class:`Category`
+        instances.
+
+        .. versionchanged:: 2.0
+
+            Now it is possible to specify :class:`Category` instances
+            in ``categories`` argument.
```
Comment (foresmac) [inline docs/ref/utils.txt]: See above

File: docs/releases/2.0.txt [hunk 4]
```diff
@@ -102,7 +102,9 @@ Minor features
 :mod:`django.contrib.syndication`
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-* ...
+* Class :class:`~django.utils.feedgenerator.Category` allows to
+  specify categories with ``scheme`` (for Atom feeds), ``domain``
```
Comment (foresmac) [inline docs/releases/2.0.txt]: Instead of "allows to specify", try "allows specifying".

**Comment by PavloKapyshin** [comment]
Thank you, @foresmac, I’ve updated pull request.

File: django/utils/feedgenerator.py [hunk 5]
```diff
@@ -109,19 +109,29 @@ def to_str(s):
         self.feed.update(kwargs)
         self.items = []
 
+    def _convert_categories(self, categories):
+        res = []
+        for cat in categories:
+            if not isinstance(cat, Category):
+                cat = force_text(cat, strings_only=True)
```
Comment (vytisb) [inline django/utils/feedgenerator.py]: Is `string_only=True` required here? If it is, it's not tested.

**Comment by PavloKapyshin** [comment]
Good point, @vytisb, thank you. I extended test feeds’ categories to check for use of `force_text`.

**Comment by PavloKapyshin** [comment]
I’ve updated pull request to keep up with master.

**Comment by timgraham** [comment]
Considering all changes are in `utils/feedgenerator.py`, what do you think about putting the tests in `tests/utils_tests/test_feedgenerator.py` rather than in the syndication tests? It might be overcomplicated to go through the request/response cycle to test this.

File: tests/syndication_tests/feeds.py [hunk 6]
```diff
@@ -53,6 +53,22 @@ class TestRss091Feed(TestRss2Feed):
...
+        feedgenerator.Category(
+            'some', label='Something', domain='http://example.com/Some'
+        ),
+        feedgenerator.Category('django', scheme='framework:Django'),
+        feedgenerator.Category(123, domain=b'some-uri', label=3),
+    )
+    item_categories = (
+        'python', 'testing', 11, 12,
+        feedgenerator.Category('simple', domain='http://example.org/Simple'),
+        feedgenerator.Category(b'creativity in tests', scheme=b'bytes here'),
```
Comment (timgraham) [inline tests/syndication_tests/feeds.py]: I'm doubtful that there's a need for bytestring handling considering the sentiment in https://code.djangoproject.com/ticket/27795. Did you have a use case in mind? Could some `force_text()` usage be removed instead?

**Comment by PavloKapyshin** [comment]
I thought about tests initially, and concluded that given this patch affects syndication and includes documentation additions related to it, it’d be beneficial to have tests for an API that is documented.

Replaced bytestrings with realistic example: translated strings. Replaced `force_text` with `str` (to keep patch focused, only where code is changed).

**Comment by PavloKapyshin** [comment]
So, @timgraham, what do you think? Should I move tests to `test_feedgenerator.py`?

**Comment by timgraham** [comment]
I think simpler tests are better. Existing tests might also be more complicated than needed. Testing something in django.utils exclusively in a a contrib app doesn't seem ideal. Some tests in syndication could be fine if there is some reason to suspect it's 'usage of FeedGenerator might affect things but testing all aspects of FeedGenerator via syndication feels more complicated than needed.

**Comment by PavloKapyshin** [comment]
@timgraham, I’ve moved and simplified tests. Patch is updated. Unfortunately, in addition to that—to make `test_atom_support` work—I had (for now) to pass `enclosures=()` to `feed.add_item`.

[75cf9b](https://github.com/django/django/commit/75cf9b5ac031feb8f94271c9906157c921a14520) introduced a regression by removing `if enclosure is None:`: when `add_item` in Atom feed does not receive `enclosures` (that is, when its value is `None`), `TypeError` is raised (due to `for enclosure in item['enclosures']:` in `add_item_elements` of `Atom1Feed`).

When aforementioned regression will be fixed, I’ll update this pull request.

**Comment by timgraham** [comment]
Thanks, are you able to fix that regression also?

**Comment by PavloKapyshin** [comment]
Sure: #8110.

**Comment by PavloKapyshin** [comment]
Thank you, @timgraham. I updated pull request.

**Comment by PavloKapyshin** [comment]
Just wanted to say I did not give up: I rebase this on top of `master` every ~10 days :)

**Comment by timgraham** [comment]
Pavlo, did you close this by accident?

**Comment by PavloKapyshin** [comment]
Tim, that was deliberate action. I think feed generation (syndication) should live in separate specialized library, that is not tied to particular web framework or ORM.

**Comment by timgraham** [comment]
Okay, so you would withdraw the ticket and advocate for removing contrib.syndication from Django (or at least not make further enhancements to it)?

**Pull request closed without merging**